crates/astria-conductor/src/sequencer.rs

```diff
@@ -0,0 +1,340 @@
+    task::{
+        JoinError,
+        JoinHandle,
+    },
+};
+use tracing::{
+    error,
+    info,
+    instrument,
+    warn,
+};
+
+use crate::{
+    client_provider::{
+        self,
+        ClientProvider,
+    },
+    executor,
+    executor::ExecutorCommand,
+};
+
+type SyncBlockStream = FuturesOrdered<Pin<Box<dyn Future<Output = (u32, SyncBlockResult)> + Send>>>;
+
+type SyncBlockResult = Result<SequencerBlockData, Error>;
+
+type SyncHeightResult = Result<(u32, Object<ClientProvider>), PoolError<client_provider::Error>>;
+
+#[derive(Debug, thiserror::Error)]
+enum Error {
+    #[error("failed requesting a client from the pool")]
+    Pool(#[from] PoolError<client_provider::Error>),
+    #[error("sequencer request failed")]
+    Request(#[from] sequencer_client::extension_trait::Error),
+}
+
+pub(crate) struct Reader {
+    /// The channel used to send messages to the executor task.
+    executor_tx: executor::Sender,
+
+    initial_sequencer_block_height: u32,
+
+    pool: Pool<ClientProvider>,
+
+    shutdown: oneshot::Receiver<()>,
+}
+
+impl Reader {
+    #[instrument(name = "driver", skip_all)]
+    pub(crate) async fn new(
+        initial_sequencer_block_height: u32,
+        pool: Pool<ClientProvider>,
+        shutdown: oneshot::Receiver<()>,
+        executor_tx: executor::Sender,
+    ) -> eyre::Result<Self> {
+        Ok(Self {
+            initial_sequencer_block_height,
+            executor_tx,
+            pool,
+            shutdown,
+        })
+    }
+
+    #[instrument(skip_all)]
+    pub(crate) async fn run_until_stopped(mut self) -> eyre::Result<()> {
+        use futures::{
+            stream::FuturesOrdered,
+            StreamExt as _,
+            TryFutureExt as _,
+        };
+        let mut new_blocks: stream::Fuse<NewBlocksStream> = subscribe_new_blocks(self.pool.clone())
+            .await
+            .wrap_err("failed to start initial new-blocks subscription")?
+            .fuse();
+
+        let mut pending_blocks = FuturesOrdered::new();
+        let latest_height = match new_blocks.next().await {
+            None => bail!("subscription to sequencer for new blocks failed immediately; bailing"),
+            Some(Err(e)) => {
+                return Err(e).wrap_err("first sequencer block returned from subscription was bad");
+            }
+            Some(Ok(block)) => {
+                let height = block.header().height.value();
+                pending_blocks.push_back(futures::future::ready(Ok(block)));
+                height
+            }
+        };
+
+        let latest_height: u32 = latest_height.try_into().wrap_err(
+            "failed converting the cometbft height to u32, but this should always work",
+        )?;
+
+        info!(
+            height.initial = self.initial_sequencer_block_height,
+            height.latest = latest_height,
+            "syncing sequencer between configured initial and latest retrieved height"
+        );
+        let mut sync_heights = {
```

Comment:
To make sure I understand correctly, we now only sync on startup? I believe in the previous version we looked over if a block came via the subscription out of order we would also start a sync on the missing blocks?

Is a sync job potentially something we can trigger via an inbound command? This would allow the executor to have an execution queue and start a sync if something was missing?

Reply:
sync on startup: Yes, that is correct.

previous version: Looking at current `main`, I don't think we had that, at least not for the sequencer part.

trigger sync via command: I think we can leave this squarely inside the `sequencer::Reader` actor so that executor need not be concerned with it: we can add logic to the sequencer reader to track the current height and require it be contiguous. If we detect a hole, we can explicitly fetch that block and pause forwarding to the executor until the hole is plugged.

Follow-up PR?

Reply:
Created an issue.